pkg/coordinator/coordinator.go

```diff
@@ -1,109 +1,154 @@
+// should be started before use.
+func NewReportCoordinator(
+	lockoutWindow time.Duration,
+	cacheClean time.Duration,
+	logs LogProvider,
+	minConfs int,
+	logger *log.Logger,
+	encoder Encoder,
+) *reportCoordinator {
+	if lockoutWindow < 1 {
+		lockoutWindow = DefaultLockoutWindow
+	}
+
+	if cacheClean < 1 {
+		cacheClean = DefaultCacheClean
+	}
+
 	c := &reportCoordinator{
 		logger:         logger,
-		registry:       r,
 		logs:           logs,
 		minConfs:       minConfs,
-		idBlocks:       util.NewCache[idBlocker](s),
+		idBlocks:       util.NewCache[idBlocker](lockoutWindow),
 		activeKeys:     util.NewCache[bool](time.Hour), // 1 hour allows the cleanup routine to clear stale data
 		idCacheCleaner: util.NewIntervalCacheCleaner[idBlocker](cacheClean),
 		cacheCleaner:   util.NewIntervalCacheCleaner[bool](cacheClean),
-		chStop:         make(chan struct{}),
+		chStop:         make(chan struct{}, 1),
+		encoder:        encoder,
 	}
 
-	runtime.SetFinalizer(c, func(srv *reportCoordinator) { srv.stop() })
-
-	c.start()
-
 	return c
 }
 
-// Filter returns a filter function that removes upkeep keys that apply to this
-// filter. Returns false if a key should be filtered out.
-func (rc *reportCoordinator) Filter() func(types.UpkeepKey) bool {
-	return func(key types.UpkeepKey) bool {
-		blockKey, id, err := key.BlockKeyAndUpkeepID()
-		if err != nil {
-			return false
-		}
-
-		// only apply filter if key id is registered in the cache
-		if bl, ok := rc.idBlocks.Get(string(id)); ok {
-			isAfter, err := blockKey.After(bl.TransmitBlockNumber)
-			if err != nil {
-				return false
-			}
+// IsPending returns true if a key should be filtered out.
+func (rc *reportCoordinator) IsPending(key ocr2keepers.UpkeepKey) (bool, error) {
+	blockKey, id, err := rc.encoder.SplitUpkeepKey(key)
+	if err != nil {
+		return true, fmt.Errorf("%w: key parse error", err)
+	}
 
-			// do not filter the key out if key block is after block in cache
-			return isAfter
+	// only apply filter if key id is registered in the cache
+	if bl, ok := rc.idBlocks.Get(string(id)); ok {
+		isAfter, err := rc.encoder.After(blockKey, bl.TransmitBlockNumber)
+		if err != nil {
+			return true, fmt.Errorf("%w: not after transmit number", err)
 		}
 
-		return true
+		// do not filter the key out if key block is after block in cache
+		return !isAfter, nil
 	}
+
+	return false, nil
 }
 
-func (rc *reportCoordinator) Accept(key types.UpkeepKey) error {
-	blockKey, id, err := key.BlockKeyAndUpkeepID()
+// Accept sets the pending status for a key
+func (rc *reportCoordinator) Accept(key ocr2keepers.UpkeepKey) error {
+	blockKey, id, err := rc.encoder.SplitUpkeepKey(key)
 	if err != nil {
 		return err
 	}
 
 	// If a key is already active then don't update filters, but also not throw errors as
 	// there might be other keys in the same report which can get accepted
-	if _, ok := rc.activeKeys.Get(key.String()); !ok {
+	// TODO: key to string again
```

Comment:
can we link a ticket please

Reply:
This was addressed in the original review here: https://github.com/smartcontractkit/ocr2keepers/pull/113#discussion_r1197163626

Reply:
so this is intentional to not have string method on a []byte? we shouldn't have the TODO then

fwiw it's not clear from this snippet that it will always be a []byte as it is an interface type. I thought the whole purpose was not to tie the logic with concrete types :upside_down_face:

Reply:
I discussed this with @ferglor and we both think it would be better to have a string interface to avoid implicit assumption of underlying type being bytes. 

@ferglor feel free to put up a PR adding the string interface

Reply:
If we do this and pass around interfaces, we break the ability of having static types. We can't have both.